src/backend/distributed/planner/shard_pruning.c

```diff
@@ -0,0 +1,1267 @@
+		{
+			ListCell *opCell = NULL;
+
+			/*
+			 * "Queue" partial pruning instances.  This is used to convert
+			 * expressions like (A AND (B OR C) AND D) into (A AND B AND D),
+			 * (A AND C AND D), with A, B, C, D being restrictions.  When the
+			 * OR is encountered, a reference to the partially built
+			 * PruningInstance (containing A at this point), is added to
+			 * context->pendingInstances once for B and once for C.  Once a
+			 * full tree-walk completed, PrunableExpressions() will complete
+			 * the pending instances, which'll now also know about restriction
+			 * D, by calling PrunableExpressionsWalker() once for B and once
+			 * for C.
+			 */
+			foreach(opCell, boolExpr->args)
+			{
+				PendingPruningInstance *instance =
+					palloc0(sizeof(PendingPruningInstance));
+
+				instance->instance = context->currentPruningInstance;
+				instance->continueAt = lfirst(opCell);
+
+				/*
+				 * Signal that this instance is not to be used for pruning on
+				 * its own.  Once the pending instance is processed, it'll be
+				 * used.
+				 */
+				instance->instance->isPartial = true;
+
+				context->pendingInstances = lappend(context->pendingInstances, instance);
+			}
+
+			return false;
+		}
+	}
+	else if (IsA(node, OpExpr))
+	{
+		OpExpr *operatorExpression = (OpExpr *) node;
+		PruningInstance *prune = context->currentPruningInstance;
+		Node *leftOperand = NULL;
+		Node *rightOperand = NULL;
+		Const *constantClause = NULL;
+		Var *varClause = NULL;
+
+		if (!prune->addedToPruningInstances)
+		{
+			context->pruningInstances = lappend(context->pruningInstances,
+												prune);
+			prune->addedToPruningInstances = true;
+		}
+
+		if (list_length(operatorExpression->args) == 2)
+		{
+			leftOperand = get_leftop((Expr *) operatorExpression);
+			rightOperand = get_rightop((Expr *) operatorExpression);
+
+			leftOperand = strip_implicit_coercions(leftOperand);
+			rightOperand = strip_implicit_coercions(rightOperand);
+
+			if (IsA(rightOperand, Const) && IsA(leftOperand, Var))
+			{
+				constantClause = (Const *) rightOperand;
+				varClause = (Var *) leftOperand;
+			}
+			else if (IsA(leftOperand, Const) && IsA(rightOperand, Var))
+			{
+				constantClause = (Const *) leftOperand;
+				varClause = (Var *) rightOperand;
+			}
+		}
+
+		if (constantClause && varClause && equal(varClause, context->partitionColumn))
+		{
+			List *btreeInterpretationList = NULL;
+			ListCell *btreeInterpretationCell = NULL;
+			bool matchedOp = false;
+
+			/*
+			 * Found a restriction on the partition column itself. Update the
+			 * current constraint with the new information.
+			 */
+
+			btreeInterpretationList =
+				get_op_btree_interpretation(operatorExpression->opno);
+			foreach(btreeInterpretationCell, btreeInterpretationList)
+			{
+				OpBtreeInterpretation *btreeInterpretation =
+					(OpBtreeInterpretation *) lfirst(btreeInterpretationCell);
+
+				switch (btreeInterpretation->strategy)
```

Comment:
Done, into two different functions for the magic col, and plain constraint cases.